packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts

```diff
@@ -0,0 +1,323 @@
+            summaryTree.tree[key] = {
+                type: api.SummaryType.Blob,
+                content: hash,
+            };
+            const fullBlobPath = path === "" ? `.app/${key}` : `${path}/${key}`;
+            this.blobTreeDedupCaches.blobShaToPath.set(hash, fullBlobPath);
+            this.blobTreeDedupCaches.pathToBlobSha.set(fullBlobPath, hash);
+        }
+
+        for (const [key, tree] of Object.entries(snapshotTree.trees)) {
+            const fullTreePath = path === "" ? key === ".protocol" ? `${key}` : ".app" : `${path}/${key}`;
+            const subtree = await this.buildCachesForDedup(tree, fullTreePath);
+            this.blobTreeDedupCaches.treesPathToTree.set(fullTreePath, subtree);
+            summaryTree.tree[key] = subtree;
+        }
+        return summaryTree;
+    }
+
+    public async writeSummaryTree(tree: api.ISummaryTree, context: ISummaryContext) {
+        // If the last proposed handle is not the proposed handle of the acked summary(could happen when the last summary get nacked),
+        // then re-initialize the caches with the backup ones.
+        if (context.proposalHandle !== this.lastSummaryProposalHandle) {
+            this.blobTreeDedupCaches = { ...this.backupBlobTreeDedupCaches };
+        }
+        this.backupBlobTreeDedupCaches = { ...this.blobTreeDedupCaches };
+
+        const { result, blobTreeDedupCachesLatest } = await this.writeSummaryTreeCore(context.ackHandle, tree, 0);
+        const id = result ? result.id : undefined;
+        if (!result || !id) {
+            throw new Error(`Failed to write summary tree`);
+        }
+        this.blobTreeDedupCaches = { ...blobTreeDedupCachesLatest };
+        this.lastSummaryProposalHandle = id;
+
+        return id;
+    }
+
+    private async writeSummaryTreeCore(
+        parentHandle: string | undefined,
+        tree: api.ISummaryTree,
+        depth: number,
+    ): Promise<{
+            result: ISnapshotResponse,
+            blobTreeDedupCachesLatest: IDedupCaches,
+    }> {
+        // This cache is associated with mapping sha to path for currently generated summary.
+        // We are building these caches from scratch as this will take care of the deleted blobs. The deleted blobs/trees will not come
+        // in these caches and then we will replace the old caches with these new caches, so that they have the correct values.
+        const blobTreeDedupCachesLatest: IDedupCaches = {
+            blobShaToPath: new Map(),
+            pathToBlobSha: new Map(),
+            treesPathToTree: new Map(),
+        };
+        const { snapshotTree, reusedBlobs, blobs } = await this.convertSummaryToSnapshotTree(
+            parentHandle,
+            tree,
+            blobTreeDedupCachesLatest,
+            0,
+            ".app",
+        );
+        const snapshot: ISnapshotRequest = {
+            entries: snapshotTree.entries!,
+            message: "app",
+            sequenceNumber: depth === 0 ? 1 : 2,
```

Comment:
I'm a bit confused by this sequenceNumber. What does it represent?
Can we not write it at all? I think this logic reminds me of what server puts into trees on reading (one tree has 0, another has real sequence number), but having 1 or 2 just does not make sense here.

Reply:
As i confirmed, server only looks at it when the snapshottype is Container. Here the type is channel. So removed it from the code.